diff -Nur traceroute-1.4a12.orig/aclocal.m4 traceroute-1.4a12/aclocal.m4
--- traceroute-1.4a12.orig/aclocal.m4	Sun Aug 15 01:50:09 1999
+++ traceroute-1.4a12/aclocal.m4	Wed Aug 15 15:57:01 2001
@@ -677,11 +677,10 @@
 AC_TRY_LINK(dnl
 ifelse([$2], [main], , dnl Avoid conflicting decl of main.
 [/* Override any gcc2 internal prototype to avoid an error.  */
-]ifelse(AC_LANG, CPLUSPLUS, [#ifdef __cplusplus
+#ifdef __cplusplus
 extern "C"
 #endif
-])dnl
-[/* We use char because int might match the return type of a gcc2
+/* We use char because int might match the return type of a gcc2
     builtin and then its argument prototype would still apply.  */
 char $2();
 ]),
