--- ./configure.in.orig	Sat Aug 11 21:35:11 2001
+++ ./configure.in	Sat Aug 11 21:35:46 2001
@@ -44,7 +44,7 @@
 	;;
 
 linux*)
-	V_INCLS="$V_INCLS -Ilinux-include"
+	V_INCLS="$V_INCLS -Ilinux-include -DHAVE_IFF_LOOPBACK -DUSE_KERNEL_ROUTING_TABLE"
 	AC_DEFINE(BYTESWAP_IP_HDR)
 	;;
 
--- ./configure.orig	Sat Aug 11 21:36:06 2001
+++ ./configure	Sat Aug 11 21:39:32 2001
@@ -761,7 +761,7 @@
   # We found a bogon in the path, so make sure we never use it.
   set dummy $ac_cv_prog_CC
   shift
-  if test $# -gt 0; then
+  if test $# -gt 0 ; then
     # We chose a different compiler from the bogus one.
     # However, it has the same basename, so the bogon will be chosen
     # first if we set CC to just the basename; use the full file name.
@@ -1789,7 +1789,7 @@
 	;;
 
 linux*)
-	V_INCLS="$V_INCLS -Ilinux-include"
+	V_INCLS="$V_INCLS -Ilinux-include -DHAVE_IFF_LOOPBACK -DUSE_KERNEL_ROUTING_TABLE"
 	cat >> confdefs.h <<\EOF
 #define BYTESWAP_IP_HDR 1
 EOF
@@ -2320,7 +2320,7 @@
 ac_more_lines=:
 ac_sed_cmds=""
 while $ac_more_lines; do
-  if test $ac_beg -gt 1; then
+  if test $ac_beg -gt 1 ; then
     sed "1,${ac_beg}d; ${ac_end}q" conftest.subs > conftest.s$ac_file
   else
     sed "${ac_end}q" conftest.subs > conftest.s$ac_file
