--- ./ifaddrlist.c.orig	Sun Aug 12 15:03:25 2001
+++ ./ifaddrlist.c	Sun Aug 12 15:05:30 2001
@@ -168,7 +168,15 @@
 			return (-1);
 		}
 		sin = (struct sockaddr_in *)&ifr.ifr_addr;
-		al->addr = sin->sin_addr.s_addr;
+		if (ioctl(fd, SIOCGIFNETMASK, (char *)&ifr) < 0) {
+			(void)sprintf(errbuf, "SIOCGIFNETMASK: %s: %s",
+				device, strerror(errno));
+			(void)close(fd);
+			return (-1);
+		}
+		
+		sin = (struct sockaddr_in *)&ifr.ifr_netmask;
+		al->mask = sin->sin_addr.s_addr;
 		al->device = strdup(device);
 		++al;
 		++nipaddr;
--- ./ifaddrlist.h.orig	Sun Aug 12 15:03:31 2001
+++ ./ifaddrlist.h	Sun Aug 12 15:05:45 2001
@@ -22,7 +22,7 @@
  */
 
 struct ifaddrlist {
-	u_int32_t addr;
+	u_int32_t mask;
 	char *device;
 };
 
